File: metodos/Doolittle.py
import numpy as np

def doolittle(A, b):
    n = len(A)
    L = np.zeros((n, n))
    U = np.zeros((n, n))
    x = np.zeros(n)

    for i in range(n):
        L[i][i] = 1.0

        for j in range(i, n):
            sum_1 = sum(L[i][k] * U[k][j] for k in range(i))
            U[i][j] = A[i][j] - sum_1

        for j in range(i + 1, n):
            sum_2 = sum(L[j][k] * U[k][i] for k in range(i))
            L[j][i] = (A[j][i] - sum_2) / U[i][i]

    # Substitucion progresiva
    y = np.zeros(n)
    for i in range(n):
        y[i] = (b[i] - np.dot(L[i, :i], y[:i])) / L[i, i]

    # sub regresiva
    for i in range(n - 1, -1, -1):
        x[i] = (y[i] - np.dot(U[i, i + 1:], x[i + 1:])) / U[i, i]

    return x


# A = np.array([[36,3,-4,5],
#               [5,-45,10,-2],
#               [6,8,57,5],
#               [2,3,-8,-42]])

# b = np.array([-20,69,96,-32])
# print(doolittle(A,b))

File: metodos/Doolittle.java
import java.util.Arrays;

public class Doolittle {
    public static double[] doolittle(double[][] A, double[] b) {
        int n = A.length;
        double[][] L = new double[n][n];
        double[][] U = new double[n][n];
        double[] x = new double[n];

        for (int i = 0; i < n; i++) {
            L[i][i] = 1.0;

            for (int j = i; j < n; j++) {
                double sum_1 = 0.0;
                for (int k = 0; k < i; k++) {
                    sum_1 += L[i][k] * U[k][j];
                }
                U[i][j] = A[i][j] - sum_1;
            }

            for (int j = i + 1; j < n; j++) {
                double sum_2 = 0.0;
                for (int k = 0; k < i; k++) {
                    sum_2 += L[j][k] * U[k][i];
                }
                L[j][i] = (A[j][i] - sum_2) / U[i][i];
            }
        }

        double[] y = new double[n];
        for (int i = 0; i < n; i++) {
            double dotProduct = 0.0;
            for (int j = 0; j < i; j++) {
                dotProduct += L[i][j] * y[j];
            }
            y[i] = (b[i] - dotProduct) / L[i][i];
        }

